client/ping.go

```diff
@@ -71,6 +74,38 @@ func parsePingResponse(cli *Client, resp serverResponse) (types.Ping, error) {
 			ControlAvailable: role == "manager",
 		}
 	}
+	if engineFeatures := resp.header.Get("Engine-Features"); engineFeatures != "" {
+		featuresMap, err := parseEngineFeaturesHeader(engineFeatures)
+		if err != nil {
+			return ping, err
+		}
+		ping.EngineFeatures = featuresMap
+	}
 	err := cli.checkResponseErr(resp)
 	return ping, errdefs.FromStatusCode(err, resp.statusCode)
 }
+
+// Prevent a malicious engine from blowing up clients by sending
+// a huge number of engine features.
+const maxNumEngineFeatures = 100
+
+func parseEngineFeaturesHeader(header string) (map[string]bool, error) {
+	if numFeatures := strings.Count(header, ",") + 1; numFeatures > maxNumEngineFeatures {
+		return nil, errors.Errorf("too many features in Engine-Features header – expected max %d, found %d", maxNumEngineFeatures, numFeatures)
+	}
+
+	features := strings.Split(header, ",")
+	featuresMap := make(map[string]bool)
+	for _, featureValue := range features {
+		parts := strings.Split(featureValue, "=")
+		if len(parts) < 2 {
+			return nil, fmt.Errorf("failed to parse Engine-Features header feature: %s – missing '='", featureValue)
+		}
+		if len(parts) > 2 {
+			return nil, fmt.Errorf("failed to parse Engine-Features header feature: %s – too many '='", featureValue)
+		}
+		featuresMap[parts[0]] = httputils.ParseBool(parts[1])
```

Comment:
Looking at the reason why we are importing httputils; this header is set by our API, so perhaps we should just consider it to be always `true` as values we return and make that part of the API contract? The only thing I could consider is if we use Features to communicate values other than booleans to be included; I know compose currently sets a Label on `docker info` that  they use for discovery;

```
Labels:
  com.docker.desktop.address=unix:///Users/thajeztah/Library/Containers/com.docker.docker/Data/docker-cli.sock
```

Reply:
Due to things like what @vvoland pointed out [here](https://github.com/moby/moby/pull/48904#discussion_r1849993028), I think it would be best to not 
> just consider it to be always true as values we return and make that part of the API contract

IMO this is just more explicit/safer, and with not that many drawbacks.